packages/jsii-pacmak/lib/npm-modules.ts

```diff
@@ -0,0 +1,115 @@
+import fs = require('fs-extra');
+import path = require('path');
+import spec = require('jsii-spec');
+import { resolveDependencyDirectory } from './util';
+
+import logging = require('../lib/logging');
+import { JsiiModule } from './packaging';
+
+
+/**
+ * Find all modules that need to be packagerd
+ *
+ * If the input list is empty, include the current directory. The result
+ * is NOT topologically sorted.
+ */
+export async function findJsiiModules(directories: string[], recurse: boolean) {
+  const ret: JsiiModule[] = [];
+  const visited = new Set<string>();
+  for (const dir of directories.length > 0 ? directories : ['.']) {
+    await visitPackage(dir, true);
+  }
+  return ret;
+
+  async function visitPackage(dir: string, isRoot: boolean) {
+    const realPath = await fs.realpath(dir);
+    if (visited.has(realPath)) { return; } // Already visited
+    visited.add(realPath);
+
+    const pkg = await fs.readJson(path.join(realPath, 'package.json'));
+    if (!pkg.jsii || !pkg.jsii.outdir || !pkg.jsii.targets) {
+      if (isRoot) {
+        throw new Error(`Invalid "jsii" section in ${realPath}. Expecting "outdir" and "targets"`);
+      } else {
+        return; // just move on, this is not a jsii package
+      }
+    }
+
+    // if --recurse is set, find dependency dirs and build them.
+    if (recurse) {
+      for (const dep of Object.keys(pkg.dependencies || {})) {
+        const depDir = resolveDependencyDirectory(realPath, dep);
+        await visitPackage(depDir, false);
+      }
+    }
+
+    // outdir is either by package.json/jsii.outdir (relative to package root) or via command line (relative to cwd)
+    const outputDirectory = pkg.jsii.outdir && path.resolve(realPath, pkg.jsii.outdir);
+    const targets = [...Object.keys(pkg.jsii.targets), 'js']; // "js" is an implicit target.
+
+    ret.push(new JsiiModule({
+      name: pkg.name,
+      moduleDirectory: realPath,
+      defaultOutputDirectory: outputDirectory,
+      availableTargets: targets
+    }));
+  }
+
+}
+
+export async function updateAllNpmIgnores(packages: JsiiModule[]) {
+  for (const pkg of packages) {
+    // updates .npmignore to exclude the output directory and include the .jsii file
+    await updateNpmIgnore(pkg.moduleDirectory, pkg.outputDirectory);
+  }
+}
+
+async function updateNpmIgnore(packageDir: string, excludeOutdir: string | undefined) {
```

Comment:
Seriously, this function is very long for the value it produces. I would be inclined to just stopping doing that.